sw/device/tests/kmac_error_conditions_test.c

```diff
@@ -0,0 +1,447 @@
+                               kKmacTestVector.mode, 0, &kKmacTestVector.key,
+                               NULL));
+
+  // Read out error information.
+  dif_kmac_error_t err;
+  uint32_t info;
+  TRY(dif_kmac_get_error(&kmac, &err, &info));
+  CHECK(err == kDifErrorKeyNotValid,
+        "Error other than kDifErrorKeyNotValid occurred.");
+
+  // Reset the KMAC FSM.
+  TRY(dif_kmac_reset(&kmac, &kmac_operation_state));
+
+  return OK_STATUS();
+}
+
+/**
+ * ErrUnexpectedModeStrength.
+ *
+ * This test checks whether the ErrUnexpectedModeStrength error code is
+ * returned by the hardware when using a wrong strength. For mode SHA3 use
+ * 128-bit strength and for SHAKE 224-bit.
+ *
+ * @return OK or error.
+ */
+status_t test_err_unexpected_mode_strength(void) {
+  LOG_INFO("Testing ErrUnexpectedModeStrength error.");
+
+  uint32_t misconfigured_strength[2] = {KMAC_CFG_SHADOWED_KSTRENGTH_VALUE_L128,
+                                        KMAC_CFG_SHADOWED_KSTRENGTH_VALUE_L224};
+  uint32_t mode[2] = {KMAC_CFG_SHADOWED_MODE_VALUE_SHA3,
+                      KMAC_CFG_SHADOWED_MODE_VALUE_SHAKE};
+
+  for (int it = 0; it < ARRAYSIZE(mode); it++) {
+    // Re-init KMAC for the test.
+    TRY(dif_kmac_init(mmio_region_from_addr(TOP_EARLGREY_KMAC_BASE_ADDR),
+                      &kmac));
+    CHECK_STATUS_OK(kmac_testutils_config(&kmac, false));
+
+    uint32_t cfg_reg =
+        mmio_region_read32(kmac.base_addr, KMAC_CFG_SHADOWED_REG_OFFSET);
+    // Misconfigure the strength.
+    cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD,
+                                     misconfigured_strength[it]);
+    cfg_reg =
+        bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD, mode[it]);
+    mmio_region_write32(kmac.base_addr, KMAC_CFG_SHADOWED_REG_OFFSET, cfg_reg);
+    mmio_region_write32(kmac.base_addr, KMAC_CFG_SHADOWED_REG_OFFSET, cfg_reg);
+
+    // Issue start command.
+    uint32_t cmd_reg =
+        bitfield_field32_write(0, KMAC_CMD_CMD_FIELD, KMAC_CMD_CMD_VALUE_START);
+    mmio_region_write32(kmac.base_addr, KMAC_CMD_REG_OFFSET, cmd_reg);
+
+    // Read out error information.
+    dif_kmac_error_t err;
+    uint32_t info;
+    TRY(dif_kmac_get_error(&kmac, &err, &info));
+    CHECK(err == kDifErrorUnexpectedModeStrength,
+          "Error other than kDifErrorUnexpectedModeStrength occurred.");
+
+    // Reset the KMAC FSM.
+    dif_kmac_operation_state_t kmac_operation_state;
+    TRY(dif_kmac_reset(&kmac, &kmac_operation_state));
+  }
+
+  return OK_STATUS();
+}
+
+/**
+ * ErrSwPushedMsgFifo.
+ *
+ * This test checks whether the ErrSwPushedMsgFifo error code is
+ * returned by the hardware when writing to the message FIFO without first
+ * sending a START command. This is different to the test description in the
+ * test plan as writing to the message FIFO during a keymanager app request is
+ * not easy to implement. However, this test still checks whether the correct
+ * error code is generated.
```

Comment:
Would this better be tested in a keymgr test?

I think we should test this somehow because it seems important to me to signal errors in this case. What is your view?

Reply:
I've rewritten the test in order to comply to the test plan.